### PR TITLE
blas/gonum: add Ztrsv

### PR DESCRIPTION
**Pull request by vladimir-ch**
I had to fix `zsame` which in turn required modifying tests for `Zher` and `Zher2`. This in turn revealed that their behavior is not as documented when `alpha == 0`. I postpone that to another PR. If you wish I can cut out the commits to a prequel PR.

### CODE REVIEW

File: blas/testblas/common.go [hunk 1]
```diff
@@ -260,7 +260,16 @@ func zsame(x, y []complex128) bool {
 	}
 	for i, v := range x {
 		w := y[i]
-		if v != w && !math.IsNaN(real(v)) && !math.IsNaN(imag(v)) && !math.IsNaN(real(w)) && !math.IsNaN(imag(w)) {
+		if math.IsNaN(real(v)) && math.IsNaN(imag(v)) && math.IsNaN(real(w)) && math.IsNaN(imag(w)) {
```
Comment (kortschak) [inline blas/testblas/common.go]: Could we not have something like 

```
func sameFloat64(a, b float64) bool {
    return a == b || (math.IsNaN(a) && math.IsNaN(b))
}
```

and then

```
for i, v := range x {
    w := y[i]
    if !sameFloat64(real(v), real(w)) || !sameFloat64(imag(v), imag(w)) {
        return false
    }
}
```

File: blas/testblas/common.go [hunk 2]
```diff
@@ -260,7 +260,16 @@ func zsame(x, y []complex128) bool {
 	}
```
Comment (kortschak) [inline blas/testblas/common.go]: Can we have an explanation in the commit message for this for why the change was necessary.

File: blas/testblas/common.go [hunk 3]
```diff
@@ -276,6 +276,28 @@ func zsame(x, y []complex128) bool {
 	return true
 }
```
Comment (kortschak) [inline blas/testblas/common.go]: This probably should have been committed before the tests, but github will never render it that way now.

File: blas/testblas/common.go [hunk 4]
```diff
@@ -276,6 +276,28 @@ func zsame(x, y []complex128) bool {
...
 
+func zequal(x, y []complex128, tol float64) bool {
+	if len(x) != len(y) {
+		return false
+	}
+	for i, v := range x {
```
Comment (kortschak) [inline blas/testblas/common.go]: Can we have a similar thing here too?

File: blas/testblas/common.go [hunk 5]
```diff
@@ -276,6 +276,28 @@ func zsame(x, y []complex128) bool {
 	return true
 }
 
+func zequal(x, y []complex128, tol float64) bool {
```
Comment (kortschak) [inline blas/testblas/common.go]: It seems that this is more `zsameApprox` rather than `zequal`.

File: blas/testblas/common.go (hunk 2 above)
Comment (vladimir-ch) [reply blas/testblas/common.go]: Done.

File: blas/testblas/common.go (hunk 3 above)
Comment (vladimir-ch) [reply blas/testblas/common.go]: I reordered the commits, so maybe now they are rendered as you wished?

File: blas/testblas/common.go (hunk 1 above)
Comment (vladimir-ch) [reply blas/testblas/common.go]: I added `sameFloat64` and then `sameComplex128` which uses it.

File: blas/testblas/common.go (hunk 4 above)
Comment (vladimir-ch) [reply blas/testblas/common.go]: Done.

File: blas/testblas/common.go (hunk 5 above)
Comment (vladimir-ch) [reply blas/testblas/common.go]: I changed the signature and the semantics of this function. I think it is better now.

**Comment by vladimir-ch** [comment]
PTAL

File: blas/testblas/common.go (hunk 3 above)
Comment (kortschak) [reply blas/testblas/common.go]: It will make the commits all pass which is the important thing. Github for some reason renders commits by creation time, not graph order.

File: blas/testblas/common.go [hunk 6]
```diff
@@ -254,13 +254,46 @@ func allPairs(x, y []int) [][2]int {
...
 		w := y[i]
-		if v != w && !math.IsNaN(real(v)) && !math.IsNaN(imag(v)) && !math.IsNaN(real(w)) && !math.IsNaN(imag(w)) {
+		if !sameComplex128(v, w) {
+			return false
+		}
+	}
+	return true
+}
+
+// zEqualApprox returns whether vectors x and y with stride inc
```
Comment (kortschak) [inline blas/testblas/common.go]: Thanks. I'm not convinced of the name, but I can't think of anything better and it doesn't face the public, so we'll leave it.

Comment (vladimir-ch) [reply blas/testblas/common.go]: What about the name worries you? I'll merge but I think there may be a chance to revisit this later. In general, the supporting functions for blas testing could be cleaned up and unified.

Comment (kortschak) [reply blas/testblas/common.go]: The name is indicating that it's testing for approximate equality, but there is no tolerance that will allow NaN == NaN, but conversely, those elements are not the elements that are being tested for approximate equality. A name that does cover it is `zApproxEqualInElementsSameOtherwise`; this is obviously not a good name though.

Comment (vladimir-ch) [reply blas/testblas/common.go]: I think that having two functions, one for stride elements and one for out-of-stride elements, would be better. I'll try to do something about it.
